### PR TITLE
feat(ws): allow closing active subscriptions

### LINKED ISSUE #4133: bug: wsLink won't close if there are running subscriptions.
### Provide environment information

 ```
 System:
    OS: Linux 5.10 Ubuntu 20.04.5 LTS (Focal Fossa)
    CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
    Memory: 6.37 GB / 12.37 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
    Yarn: 1.22.15 - ~/.yarn/bin/yarn
    npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
  Browsers:
    Chrome: 109.0.5414.119
  npmPackages:
    @trpc/client: ^10.8.1 => 10.8.1 
    @trpc/react-query: ^10.8.1 => 10.8.1 
    @trpc/server: ^10.8.1 => 10.8.1 
    typescript: ^4.8.3 => 4.9.4 
```

### Describe the bug

Explicitly calling `wsClient.close()` doesnt close the link if there are running subscriptions.
Calling close should presumably unsubscribe all running subscriptions and close websocket.
As of now the client code must keep track of all running subscriptions and close them one by one in addition to closing the wsClient.

### Link to reproduction

https://stackblitz.com/edit/github-6nnmqf?file=src/client.ts&view=editor

### To reproduce

```ts
import {
  createTRPCProxyClient,
  createWSClient,
  httpLink,
  splitLink,
  wsLink,
} from '@trpc/client';
import AbortController from 'abort-controller';
import fetch from 'node-fetch';
import ws from 'ws';
import type { AppRouter } from './server';

// polyfill fetch & websocket
const globalAny = global as any;
globalAny.AbortController = AbortController;
globalAny.fetch = fetch;
globalAny.WebSocket = ws;

const wsClient = createWSClient({
  url: `ws://localhost:2022`,
});
const trpc = createTRPCProxyClient<AppRouter>({
  links: [
    // call subscriptions through websockets and the rest over http
    splitLink({
      condition(op) {
        return op.type === 'subscription';
      },
      true: wsLink({
        client: wsClient,
      }),
      false: httpLink({
        url: `http://localhost:2022`,
      }),
    }),
  ],
});

async function main() {
  const helloResponse = await trpc.greeting.hello.query({
    name: 'world',
  });

  console.log('helloResponse', helloResponse);

  const createPostRes = await trpc.post.createPost.mutate({
    title: 'hello world',
    text: 'check out https://tRPC.io',
  });
  console.log('createPostResponse', createPostRes);

  let count = 0;
  await new Promise<void>((resolve) => {
    const subscription = trpc.post.randomNumber.subscribe(undefined, {
      onData(data) {
        // ^ note that `data` here is inferred
        console.log('received', data);
        count++;
        if (count > 20) {
          subscription.unsubscribe();
        }
      },
      onError(err) {
        console.error('error', err);
      },
      onStarted(){
        resolve();
      }
    });
  });
  console.log('gonna close wsClient');
  wsClient.close();
}

main();
```

### Additional information

_No response_

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

### LINKED ISSUE #4135: Closing subscription from server results in client throwing "Operation ended Prematurely"
### Provide environment information

```
System:
   OS: Linux 5.10 Ubuntu 20.04.5 LTS (Focal Fossa)
   CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
   Memory: 6.37 GB / 12.37 GB
   Container: Yes
   Shell: 5.0.17 - /bin/bash
 Binaries:
   Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
   Yarn: 1.22.15 - ~/.yarn/bin/yarn
   npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
 Browsers:
   Chrome: 109.0.5414.119
 npmPackages:
   @trpc/client: ^10.8.1 => 10.8.1 
   @trpc/react-query: ^10.8.1 => 10.8.1 
   @trpc/server: ^10.8.1 => 10.8.1 
   typescript: ^4.8.3 => 4.9.4 
```

### Describe the bug

Closing subscription (calling the observer's `complete` function) from server results in client throwing "Operation ended Prematurely".
Also rather well described here: https://github.com/trpc/trpc/discussions/4060

### Link to reproduction

https://stackblitz.com/edit/github-6nnmqf-bkqqby?file=src/client.ts&view=editor

### To reproduce

```ts
// Somewhere earlier, outside the subscription route:
let onData: ((data: string) => void) | undefined = undefined;
let onComplete: (() => void) | undefined = undefined;

// Inside the subscription route
return observable<string>((emit) => {
  onData = emit.next;
  onComplete = emit.complete;

  return () => {
    onData = undefined;
    onComplete = undefined;
  };
});
    
// Later when server wants to close connection
onComplete?.();
```

### Additional information

This is happening because the wsClient (actually the wsLink code) only accepts calls to the complete callback if it originates from the client (the userland code using the TRPC subscription, that is). This is achieved by using a `isdone` boolean which is only set to true when userland calls `unsubscribe`.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

### PR DESCRIPTION
**Pull request by Dealerpriest**
Closes #4133
Closes #4135 

## 🎯 Changes
Don't throw "Operation ended prematurely" when calling `complete` from outside `unsubscribeFunc`.
wsClient closes all active subscriptions when wsClient is closed allowing the wsClient to actually close the websocket.

## ✅ Checklist

- [ ] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

**Review by KATT — DISMISSED** [review]
LGTM :)

Feel free to request another review once build is passing

**Comment by sachinraja** [comment]
looks like this breaks a test, could you take a look @Dealerpriest?

**Comment by Dealerpriest** [comment]
I'll try to get some time to have a look the coming week!

**Comment by bautistaaa** [comment]
> @sachinraja I have made some updates:
> 
> * Changed so that the client only calls `complete` on pending subscriptions and not all  `pendingRequests`. This will allow a graceful closing, waiting for slow mutations and queries.
> * Updated a test where the server closes the subscription to no longer expect `onError` to be called.
> 
> Unfortunately there is still one test I don't know how to fix. It's the `createContext throws` that expects `createContext` to be called one time, but it gets called twice. I'm not sure why that happens. I get the feeling its inside the testhelper but not sure. Any ideas are much welcomed!

@juliusmarminge not sure if related but i remember you were dealing with something called twice(i forget the context). anyhow, just poking you just in case it was related

**Comment by juliusmarminge** [comment]
> > @sachinraja I have made some updates:
> 
> > 
> 
> > * Changed so that the client only calls `complete` on pending subscriptions and not all  `pendingRequests`. This will allow a graceful closing, waiting for slow mutations and queries.
> 
> > * Updated a test where the server closes the subscription to no longer expect `onError` to be called.
> 
> > 
> 
> > Unfortunately there is still one test I don't know how to fix. It's the `createContext throws` that expects `createContext` to be called one time, but it gets called twice. I'm not sure why that happens. I get the feeling its inside the testhelper but not sure. Any ideas are much welcomed!
> 
> 
> 
> @juliusmarminge not sure if related but i remember you were dealing with something called twice(i forget the context). anyhow, just poking you just in case it was related

That was a flaky one and we re-run those like 5 times in CI so if its failing everytime i dont think thats it

File: packages/client/src/links/wsLink.ts [hunk 1]
```diff
@@ -120,6 +120,14 @@ export function createWSClient(opts: WebSocketClientOptions) {
     }
   }
 
+  function closeActiveSubscriptions() {
+    Object.values(pendingRequests).forEach((req) => {
+      if (req.type === 'subscription') {
+        req.callbacks.complete();
+      }
+    });
+  }
```
Comment (KATT) [inline packages/client/src/links/wsLink.ts:129]: Could you add a test for this?

Comment (KATT) [reply packages/client/src/links/wsLink.ts:129]: I.e. making sure that if you have an active sub & call `.close()` it should just close

Comment (Dealerpriest) [reply packages/client/src/links/wsLink.ts:129]: Will give it a shot! What's the purpose/differentiation between the interop tests and the usual ones? At first glance I got the impression they are just a duplication ?

Comment (Dealerpriest) [reply packages/client/src/links/wsLink.ts:129]: Is there a reason for not merging? I have added/edited tests to account for the new functionality. I think it should be correct😊👍

**Review by KATT — APPROVED** [review]

**Pull request merged**